widget组件，align属性增加ONCE_WINDOW_RESIZED选项

PPpro opened the pull request:
Re: #2395 

Changelog:
 * 

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!

- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.

- To official teams:
  - [ ] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
  - [ ] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)
  - [ ] Make sure any runtime log information in `cc.log` , `cc.error` or `new Error('')` has been moved into `DebugInfos.js` with an ID, and use `cc.logID(id)` or `new Error(cc._getErrorID(id))` instead.

-->

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -33,6 +33,12 @@ var RIGHT   = 1 << 5;
+var AlignMode = cc.Enum({
+    ALIGN_ONCE: 0,
+    ALWAYS: 1,
+    ONCE_WINDOW_RESIZED: 2,
```
ALIGN_ONCE -> ONCE
ONCE_WINDOW_RESIZED -> ON_WINDOW_RESIZED
另外顺序调整下，ALWAYS 放到最后

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +488,20 @@ var Widget = cc.Class({
+        },
+
+        alignMode: {
+           default: WidgetManager.AlignMode.ALIGN_ONCE,
```
默认值请用 ON_WINDOW_RESIZED

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +467,31 @@ var widgetManager = cc._widgetManager = module.exports = {
+        }
+    },
+    updateAlignment: updateAlignment,
+    AlignMode: AlignMode,
```
请参考 CCSprite.js 中的
`var SpriteType = cc.Scale9Sprite.RenderingType;` 以及
`Type: SpriteType,`
，将 AlignMode 的 API 定义在 CCWidget 内，并且暴露到 cc.Widget.AlignMode

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +488,20 @@ var Widget = cc.Class({
         isAlignOnce: {
-            default: true,
+            get: function () {
+                return this._isAlignOnce;
```
return this.alignMode === ONCE

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +488,20 @@ var Widget = cc.Class({
+            },
             tooltip: CC_DEV && 'i18n:COMPONENT.widget.align_once',
-            displayName: "AlignOnce"
+            displayName: "AlignOnce",
```
移除 displayName 和 tooltip

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +488,20 @@ var Widget = cc.Class({
+                return this._isAlignOnce;
+            },
+            set: function (value) {
+                this._isAlignOnce = value;
```
this.alignMode = value ? ONCE : ALWAYS

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -307,7 +313,9 @@ function refreshScene () {
                 for (iterator.i = 0; iterator.i < activeWidgets.length; ++iterator.i) {
                     widget = activeWidgets[iterator.i];
-                    align(widget.node, widget);
+                    if (widget.alignMode === AlignMode.ALWAYS){
```
这里不用加这个判断，因为在 activeWidgets 里的 widget 总会是需要 align 的

PPpro replied:
ON_WINDOW_RESIZED 的widget 应该也算是activeWidgets 哦

jareguo replied:
这个部分待会我会再回复

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -234,7 +240,7 @@ function visitNode (node) {
         }
         align(node, widget);
-        if ((!CC_EDITOR || animationState.animatedSinceLastFrame) && widget.isAlignOnce) {
+        if ((!CC_EDITOR || animationState.animatedSinceLastFrame) && widget.alignMode === AlignMode.ALIGN_ONCE) {
```
widget.alignMode === AlignMode.ALIGN_ONCE 改为
widget.alignMode !== AlignMode.ALWAY，把 ON_WINDOW_RESIZED 也包含进来

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -435,6 +443,13 @@ var widgetManager = cc._widgetManager = module.exports = {
+        }
+        else {  // desktop browser
+            cc.eventManager.addCustomListener('canvas-resize', this.onResized);
+        }
```
@pandamicro 有没有跨平台的监听 canvas resize 的接口？

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +467,31 @@ var widgetManager = cc._widgetManager = module.exports = {
+    onResized() {
+        var scene = cc.director.getScene();
+        if (scene) {
+            widgetManager.isAligning = true;
```
这里不需要在 resize 时立刻进行 align。而是应该更新 widget 的 enabled 状态，待 refreshScene 在下一帧调用时自然就会刷新到。

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -33,12 +33,6 @@ var RIGHT   = 1 << 5;
-    ALIGN_ONCE: 0,
-    ALWAYS: 1,
-    ONCE_WINDOW_RESIZED: 2,
-});
```
抱歉，我前面没说清楚，AlignMode 的枚举仍然在这里。CCWidget 中通过 require 去获得。
但是要将 AlignMode API 注释写在 CCWidget 内，以便生成 API 文档。对用户来说，这个 API 需要暴露到 cc.Widget.AlignMode。对引擎来说，用原有的方式访问就好。

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -530,13 +534,16 @@ var Widget = cc.Class({
     },
 
+    statics: {
+        AlignMode: AlignMode,
```
👍

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +494,18 @@ var Widget = cc.Class({
+                return this.alignMode === AlignMode.ONCE;
+            },
+            set: function (value) {
+                this.alignMode = value? AlignMode.ONCE: AlignMode.ALWAYS;
```
三元运算符左右两边请都保留一个空格（刚刚已补充到编码规范）

jareguo commented on cocos2d/core/components/CCWidget.js:
```diff
@@ -488,9 +494,18 @@ var Widget = cc.Class({
+
+        alignMode: {
+           default: AlignMode.ON_WINDOW_RESIZED,
+           type: AlignMode,
```
请增加帮忙 `tooltip: CC_DEV && 'i18n:COMPONENT.widget.align_mode',`
align_mode 我会另外定义

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -435,6 +435,13 @@ var widgetManager = cc._widgetManager = module.exports = {
+        }
+        else {  // desktop browser
+            cc.eventManager.addCustomListener('canvas-resize', this.onResized.bind(this));
+        }
```
这里漏了 JSB 的情形

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
         }
     },
-    updateAlignment: updateAlignment
+    onResized() {
```
`()` 左右两边请保留一个空格（刚刚已补充到编码规范）

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
+            this.visitNodeOnResized(scene);
+        }
+    },
+    visitNodeOnResized(node){
```
这个方法名未能准确表达该方法的实现意图，建议改叫 “refreshWidgetOnResized”

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
+        }
+    },
+    visitNodeOnResized(node){
+        var widget = !(node instanceof cc.Scene) && node.getComponent(cc.Widget);
```
var widget = cc.Node.isNode(node) && node.getComponent(cc.Widget);

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
+    visitNodeOnResized(node){
+        var widget = !(node instanceof cc.Scene) && node.getComponent(cc.Widget);
+        if (widget) {
+            if (CC_DEV) {
```
这里不需要重复做这个检测了，这个检测之前第一次对齐时已经进行过了

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
+                    }
+                }
+            }
+            if ((!CC_EDITOR || animationState.animatedSinceLastFrame) && widget.alignMode === cc.Widget.AlignMode.ON_WINDOW_RESIZED) {
```
不管是否动画编辑模式都需要重新激活，因此 `(!CC_EDITOR || animationState.animatedSinceLastFrame) && ` 去掉

jareguo commented on cocos2d/core/base-ui/CCWidgetManager.js:
```diff
@@ -452,7 +459,38 @@ var widgetManager = cc._widgetManager = module.exports = {
+        var children = node._children;
+        for (var i = 0; i < children.length; i++) {
+            var child = children[i];
+            if (child._active) {
```
就算 child 未激活，也要重新 enable 它们的 widget，因为目前 child 未激活，所以 enable widget 实际上不会有任何改变。不过当下次 child 被重新激活时，里面的 widget 就能够再生效一次，这应该更加符合用户的预期。

PPpro committed:
add align ONCE_WINDOW_RESIZED to widget
some changes in widget alignMode


defined AlignMode in CCWidgetManager.js


inspector 

edit

PPpro committed:
deal with the compatibility of alignMode

PPpro committed:
delete _isUserDisable
